jormungandr/src/blockchain/index.rs

```diff
@@ -0,0 +1,131 @@
+use crate::blockchain::storage::{Storage, StorageError};
+use chain_impl_mockchain::header::{BlockDate, ChainLength, Header, HeaderId};
+use std::collections::BTreeMap;
+use thiserror::Error;
+use tokio::{prelude::*, sync::lock::Lock};
+
+#[derive(Clone)]
+pub struct Index {
+    inner: Lock<IndexInternal>,
+}
+
+#[derive(Default)]
+struct IndexInternal {
+    chain_length_index: BTreeMap<ChainLength, HeaderId>,
+    block_date_index: BTreeMap<BlockDate, HeaderId>,
+}
```

Comment:
The this index is only valid in the context of a given state of the blockchain. If the `tip` changes the state of the blockchain will chain as we are moving to a new branch of the blockchain. This will mean some blocks indexed won't be in the blockchain anymore.

I would suggest using the HAMT instead here, and instead of a `Lock<InternalIndex>` to have a `Multiverse` of Index.

Reply:
Thanks for the suggestion I will look into that. So the idea basically is to have index for every branch and to select the appropriate index by the tip id?

Reply:
no need for the multiverse actually here. Using the HAMT is enough here. Every time you add value you create a new state of the index that you attach in the Branch when the branch is updated.